packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -71,104 +71,247 @@ export function matches(element: Element, selector: AST): boolean {
+      if (argument === undefined) {
+        return false;
+      }
+
+      const indexOfEqual = argument.indexOf('=');
+      let namePattern, valuePattern;
+      if (indexOfEqual === -1) {
+        namePattern = argument;
+      } else {
+        namePattern = argument.slice(0, indexOfEqual);
+        valuePattern = argument.slice(indexOfEqual + 1);
+      }
+
+      let value;
+      if (namePattern.startsWith('/') && namePattern.endsWith('/')) {
+        // matching attribute name by regex
+        const regex = new RegExp(namePattern.slice(1, -1));
+        const attribute = [...element.attributes].find((attr) => regex.test(attr.name));
+        if (attribute === undefined) {
+          return false;
+        }
+        value = attribute.value;
+      } else {
+        // matching attribute name by string
+        value = element.getAttribute(namePattern);
+        // null means the attribute is not present
+        if (value === null) {
+          return false;
+        }
+      }
+
+      // early exit if no value pattern is provided
+      if (!valuePattern) {
+        return true;
+      }
+
+      // wrapping quotes are optional
+      if (
+        (valuePattern.startsWith('"') && valuePattern.endsWith('"')) ||
+        (valuePattern.startsWith("'") && valuePattern.endsWith("'"))
+      ) {
+        valuePattern = valuePattern.slice(1, -1);
+      }
+
+      if (valuePattern.startsWith('/') && valuePattern.endsWith('/')) {
+        // matching value by regex
+        const regex = new RegExp(valuePattern.slice(1, -1));
+        return regex.test(value);
+      } else {
+        // matching value by string
+        return value === valuePattern;
+      }
+    } else if (selector.name === 'upward') {
+      // :upward is handled in querySelectorAll
+      return false;
     }
   }
 
   return false;
 }
 
-export function querySelectorAll(element: Element, selector: AST): Element[] {
-  const elements: Element[] = [];
+function findAncestorByDistance(element: Element, distance: number): Element | null {
+  // sanity check, since the distance is provided as part of filter
+  if (distance <= 0 || distance >= 256) {
+    return null;
+  }
+  let ancestor: Element | null = element;
+  for (let i = 0; i < distance; i++) {
+    ancestor = ancestor.parentElement;
+    if (ancestor === null) {
+      return null;
+    }
+  }
+  return ancestor;
+}
 
-  if (
-    selector.type === 'id' ||
-    selector.type === 'class' ||
-    selector.type === 'type' ||
-    selector.type === 'attribute'
-  ) {
-    elements.push(...element.querySelectorAll(selector.content));
-  } else if (selector.type === 'list') {
-    for (const subSelector of selector.list) {
-      elements.push(...querySelectorAll(element, subSelector));
+function findAncestorBySelector(element: Element, selector: string): Element | null {
+  let ancestor: Element | null = element.parentElement;
+  while (ancestor !== null) {
+    if (ancestor.matches && ancestor.matches(selector)) {
```

Comment:
Is `HTMLElement.prototype.matches` optional by any case? https://developer.mozilla.org/en-US/docs/Web/API/Element/matches